File: 0063-unique-paths-ii/0063-unique-paths-ii.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
    //    1 1 % 1
    //   [1 1 1 *]
        int m = obstacleGrid.length, n = obstacleGrid[0].length;
        int[] dp = new int[n];
        dp[n - 1] = obstacleGrid[m - 1][n - 1] == 1 ? 0 : 1;

        for (int i = m - 1; i >= 0; i--) {
            for (int j = n - 1; j >= 0; j--) {
                if (obstacleGrid[i][j] == 1) {
                    dp[j] = 0;
                    continue;
                }
                dp[j] += (j < n - 1 ? dp[j + 1] : 0);
            }
        }

        return dp[0];
    }
}

File: 0063-unique-paths-ii/0063-unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        ROWS = len(obstacleGrid)
        COLS = len(obstacleGrid[0])
        dp = {}
        @cache
        def dfs(r, c):
            if r < 0 or c < 0 or r >= ROWS or c >= COLS or obstacleGrid[r][c] == 1:
                return 0
            if r == ROWS - 1 and c == COLS - 1:
                return 1
            # if (r,c) in dp:
            #     return dp[(r,c)]
            dp[(r,c)] = dfs(r+1, c) + dfs(r, c+1)
            return dp[(r,c)]
        return dfs(0,0)